lib/src/repo.rs

```diff
@@ -1192,12 +1192,15 @@ impl IdIndex {
         Self::default()
     }
 
+    /// Inserts a bytes key to the index.
     pub fn insert(&mut self, key: &[u8], value: IdIndexValue) -> Option<IdIndexValue> {
         self.0.insert(key.to_vec(), value)
     }
 
-    /// This function returns the shortest length of a prefix of `key` that
-    /// disambiguates it from every other key in the index.
+    /// This function returns the shortest length of a prefix of `key` in hex
```

Comment:
Now the input is in bytes and the output is in a number of half-bytes. You should, at least, make a comment about that. You could also try to create a separate type form `[u8]` for "a sequence of hexadecimal characters"  and a separate type from `usize` for "a number of hexadecimal characters". (If you don't, I might try to do something like that later)

I'm guessing you are trying to optimize away all conversions between sequences of bytes and sequences of hexadecimals. IMO, it's not worth the extra complexity here, but that might be a controversial opinion.

Reply:
I re-read the commit description:

> This helps us to migrate change_id index to ReadonlyIndex.

Do you mean "commit id" index? If so, I'm not sure this change to the change id index is necessary for that (though I may have missed something).

Or do you have short-term plans to move the change id index to ReadonlyIndex as well? In that case, as an intermediate step, this seems totally sensible.

Reply:
> IMO, it's not worth the extra complexity here

According to https://github.com/martinvonz/jj/pull/1041#issuecomment-1399225876, it saves ~700ms/~15%. That's quite significant, and I don't think this is that much more complex.

Reply:
OK, I'll defer to Martin then. Please document the fact that the output is in half-bytes. Also, fix the typo in the commit description if that's indeed a typo.

Other than that, I leave to @yuja the decision of whether to do anything else to make things clearer and how. If you don't, I might look into it later, but possibly not soon.

Overall, I now think that better integrating the IdIndex with the normal index is a clear win. Thank you for doing it!

Reply:
> Now the input is in bytes and the output is in a number of half-bytes. You should make a comment about that.

Yep, that's described in `shortest_unique_prefix_len()` doc comment.

> You could also try to create a separate type form [u8] for "a sequence of hexadecimal characters"

Maybe we can add an abstract `CommitId`/`ChangeId` type, but I'm gonna try in this PRs.

> I'm guessing you are trying to optimize away all conversions between sequences of bytes and sequences of hexadecimals. IMO, it's not worth the extra complexity here, but that might be a controversial opinion.

Kind of. IMO, it's not worth converting bytes to ASCII hex (because all ids are stored in bytes.)

> Do you mean "commit id" index?

Correct, will fix the commit description.

Reply:
> Yep, that's described in shortest_unique_prefix_len() doc comment.

Some minor nits to the doc comment:

Replace "in hex" with "as a number of hexadecimal digits."

Add an extra paragraph right after: "The length of the shortest prefix in bytes would be half that, rounded up".

Replace "The given key must be in bytes" with "The given key must be provided as bytes, not as hexadecimal digits"

Reply:
Thanks for the suggestion. I've inserted the following paragraph:
```
The given `key` must be provided as bytes, not as ASCII hexadecimal
digits. The length to be returned is a number of hexadecimal digits.
```

> Add an extra paragraph right after: "The length of the shortest prefix in bytes would be half that, rounded up".

It's unlikely the caller want a length in bytes, so I skipped it.